core/src/main/java/org/infinispan/topology/ClusterCacheStatus.java

```diff
@@ -571,22 +598,67 @@ public CacheStatusResponse doJoin(Address joiner, CacheJoinInfo joinInfo) throws
          }
-         if (isFirstMember) {
-            // This node was the first to join. We need to install the initial CH
-            CacheTopology initialTopology = createInitialCacheTopology();
+         if (status == ComponentStatus.INSTANTIATED) {
+            if (persistentState.isPresent()) {
+               if (trace) log.tracef("Node %s joining. Attempting to reform previous cluster", joiner);
+               // We can only allow this to proceed if we have a complete cluster
+               CacheTopology topology = restoreCacheTopology(persistentState.get(), joinInfo.getPersistentStateChecksum());
+               if (topology != null) {
+                  // Change our status
+                  status = ComponentStatus.RUNNING;
+                  clusterTopologyManager.broadcastTopologyUpdate(cacheName, topology, availabilityMode, isTotalOrder(), isDistributed());
+                  return new CacheStatusResponse(null, currentTopology, stableTopology, availabilityMode);
+               }
+            } else {
+               if (isFirstMember) {
+                  // This node was the first to join. We need to install the initial CH
+                  CacheTopology initialTopology = createInitialCacheTopology();
+
+                  // Change our status
+                  status = ComponentStatus.RUNNING;
 
-            // Don't need to broadcast the initial CH update, just return the cache topology to the joiner
-            // But we do need to broadcast the initial topology as the stable topology
-            clusterTopologyManager.broadcastStableTopologyUpdate(cacheName, initialTopology, isTotalOrder(), isDistributed());
+                  // Don't need to broadcast the initial CH update, just return the cache topology to the joiner
+                  // But we do need to broadcast the initial topology as the stable topology
+                  clusterTopologyManager.broadcastStableTopologyUpdate(cacheName, initialTopology, isTotalOrder(), isDistributed());
+               }
+            }
          }
-         topologyBeforeRebalance = getCurrentTopology();
 
-         availabilityStrategy.onJoin(this, joiner);
+         topologyBeforeRebalance = getCurrentTopology();
+         // Only trigger availability strategy if we have a topology installed
+         if (topologyBeforeRebalance != null)
+            availabilityStrategy.onJoin(this, joiner);
       }
 
       return new CacheStatusResponse(null, topologyBeforeRebalance, stableTopology, availabilityMode);
    }
 
+   protected CacheTopology restoreCacheTopology(ScopedPersistentState state, Optional<Integer> persistentStateHash) {
+      if (trace) log.tracef("Attempting to restore CH for cache %s", cacheName);
+
+      try {
+         ConsistentHash persistedCH = joinInfo.getConsistentHashFactory().fromPersistentState(state);
+
+         if (!getExpectedMembers().containsAll(persistedCH.getMembers())) {
+            throw new IllegalStateException();
```

Comment:
ConsistenthHashFactory.fromPersistentState(state) also throws an IllegalStateException() if it can't reinstate the CH. Mostly this happens because a "PersistentUUID" is not available in the jgroups address cache (i.e. a node hasn't joined the jgroups view yet). So "symmetry" is my driver here :)